framework/src/media/HttpInputDataSource.cpp

```diff
@@ -0,0 +1,250 @@
+/* ****************************************************************
+ *
+ * Copyright 2018 Samsung Electronics All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************/
+
+#include <tinyara/config.h>
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <debug.h>
+#include <unistd.h>
+#include <assert.h>
+#include <media/HttpInputDataSource.h>
+#include <chrono>
+
+#include "utils/MediaUtils.h"
+#include "HttpStream.h"
+#include "StreamBuffer.h"
+#include "StreamBufferReader.h"
+#include "StreamBufferWriter.h"
+
+namespace media {
+namespace stream {
+
+// Content-Type for audio stream
+static const std::string TAG_CONTENT_TYPE = "Content-Type:";
+static const std::string AAC_MIME_TYPE = "audio/aac";
+static const std::string AACP_MIME_TYPE = "audio/aacp";
+static const std::string MPEG_MIME_TYPE = "audio/mpeg";
+static const std::string OPUS_MIME_TYPE = "audio/opus";
+
+static const std::chrono::seconds WAIT_HEADER_TIMEOUT = std::chrono::seconds(3);
+static const std::chrono::seconds WAIT_DATA_TIMEOUT = std::chrono::seconds(3);
+
+HttpInputDataSource::HttpInputDataSource(const std::string &url)
+	: InputDataSource(), mUrl(url)
+{
+	medvdbg("url: %s\n", mUrl.c_str());
+}
+
+HttpInputDataSource::HttpInputDataSource(const HttpInputDataSource &source)
+	: InputDataSource(source), mUrl(source.mUrl)
+{
+}
+
+HttpInputDataSource &HttpInputDataSource::operator=(const HttpInputDataSource &source)
+{
+	InputDataSource::operator=(source);
+	return *this;
+}
+
+bool HttpInputDataSource::open()
```

Comment:
open called from preparePlayer, but I think below should be added. (please consider my less understanding regarding libCurl)
1. During open, probably this will try to connect. when connection established, it should download until buffer is reached to threshold to prevent underrun.
2. So preparing by http streaming will takes more time than file or buffer based one.
    In this case we need prepareAsync.
3. isPrepare can be used check state is ready to play or not.

Reply:
I added prepareAsync() API and an observer interface to notify preparing result, please review.